### PR TITLE
XW-4528 | Fetching ab test and instantGlobals asynchronously

### PR DESCRIPTION
**Pull request by bkoval**
@Wikia/x-wing

### CODE REVIEW

File: app/inline-scripts/instant-globals.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+				document.addEventListener('instantGlobalsLoaded', () => {
+					gettersQueue.forEach((getter) => {
+						getter.callback(window.Wikia.InstantGlobals[getter.key]);
+					});
+
+					gettersQueue = [];
+				});
+			}
+
+			isListening = true;
```
Comment (latata) [inline app/inline-scripts/instant-globals.js]: `isListening` can set inside `if`

File: app/inline-scripts/instant-globals.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+		} else {
+			gettersQueue.push({key, callback});
+
+			if (!isListening) {
+				document.addEventListener('instantGlobalsLoaded', () => {
+					gettersQueue.forEach((getter) => {
+						getter.callback(window.Wikia.InstantGlobals[getter.key]);
+					});
+
+					gettersQueue = [];
```
Comment (hakubo) [inline app/inline-scripts/instant-globals.js]: you could clean up after this
document.removeEventListener

File: app/inline-scripts/instant-globals.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	};
+
+	window.getInstantGlobal = function (key, callback) {
+		if (window.Wikia && window.Wikia.InstantGlobals) {
+			callback(window.Wikia.InstantGlobals[key]);
+		} else {
+			gettersQueue.push({key, callback});
+
+			if (!isListening) {
+				document.addEventListener('instantGlobalsLoaded', () => {
```
Comment (latata) [inline app/inline-scripts/instant-globals.js]: do not use arrow function here, object shortcuts

File: app/inline-scripts/instant-globals.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+(function () {
```
Comment (hakubo) [inline app/inline-scripts/instant-globals.js:1]: is this file compiled? or should we only use es5 here?

**Review by latata — APPROVED** [review]

**Comment by bkoval** [comment]
@Wikia/adeng could you take a look at this? Do you see any potential pitfalls in moving this ad-related code into async callbacks?

File: app/inline-scripts/instant-globals.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	};
+
+	window.getInstantGlobal = function (key, callback) {
+		function onInstantGlobalsLoaded() {
+			gettersQueue.forEach((getter) => {
+				getter.callback(window.Wikia.InstantGlobals[getter.key]);
+			});
+
+			gettersQueue = [];
+			document.removeEventListener('instantGlobalsLoaded');
```
Comment (kvas-damian) [inline app/inline-scripts/instant-globals.js]: You could use `document.addEventListener('instantGlobalsLoaded', onInstantGlobalsLoaded, {once: true});` instead

File: app/inline-scripts/instant-globals.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+		}
+
+		if (window.Wikia && window.Wikia.InstantGlobals) {
+			callback(window.Wikia.InstantGlobals[key]);
```
Comment (kvas-damian) [inline app/inline-scripts/instant-globals.js]: Maybe instead of a callback we could return a promise?

File: app/index.html [hunk 7]
```diff
@@ -4,7 +4,7 @@
 {{content-for 'head'}}
 {{content-for 'fastboot-inline-scripts-head'}}
 {{content-for 'fastboot-inline-scripts-head-tracking'}}
-<script src="/mobile-wiki/assets/app.js" async></script>
```
Comment (hakubo) [inline app/index.html]: please don't

**Review by hakubo — DISMISSED** [review]

**Comment by kvas-damian** [comment]
Please make sure that ads code is loaded after InstantGlobals are avialable: https://github.com/Wikia/mobile-wiki/blob/f5d6e917229dbbb834f5423025db8d8cfb56ea94/app/routes/application.js#L81

**Review by kvas-damian — CHANGES_REQUESTED** [review]
Please make sure `mercury_ads_js` is loaded after instantGlobals are available.

File: app/index.html (hunk 7 above)
Comment (bkoval) [reply app/index.html]: :(

**Comment by kvas-damian** [comment]
https://github.com/Wikia/mobile-wiki/pull/383/files#diff-2168f026d657db1526ae92392f844a3fR81
Is now unused

**Comment by kvas-damian** [comment]
what happens if instantGlobals request fails?

File: app/inline-scripts/instant-globals.js (hunk 4 above)
Comment (bkoval) [reply app/inline-scripts/instant-globals.js:1]: ES5-only

**Review by kvas-damian — APPROVED** [review]

**Pull request merged**